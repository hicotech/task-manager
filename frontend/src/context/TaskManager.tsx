import { createContext, useContext, useReducer } from 'react';
import { tasksManagerReducer } from './reducers/taskManager';
import type {
    TaskManagerContext,
    TaskManagerProvider,
    TaskManagerState,
} from './types/taskManager';

// Initial Task Manager context structure
const initialState: TaskManagerState = {
    tasks: [],
    completed: [],
};

// Create Task Manager context with initial data
const TaskManagerContext = createContext<TaskManagerContext>({
    state: initialState,
    dispatch: () => {
        throw new Error('Hook useTasksManager used outside of the Provider');
    },
});

/**
 * Task Manager Provider Component
 *
 * Provides access to Task Manager state object and dispatch function to any nested child components through useTasksManager hook
 * Allows nested child components to access the state and dispatch actions to modify the state
 *
 * @param children ReactNode children
 * @returns TaskManagerContext.Provider
 */
export const TaskManager = ({ children }: TaskManagerProvider) => {
    const [state, dispatch] = useReducer(tasksManagerReducer, initialState);

    return (
        <TaskManagerContext.Provider value={{ state, dispatch }}>
            {children}
        </TaskManagerContext.Provider>
    );
};

/**
 * Task Manager Hook for accessing Task Manager Context
 *
 * @returns Current state object and dispatch function
 */
export const useTasksManager = () => {
    const context = useContext(TaskManagerContext);

    if (!context) {
        throw new Error('Hook useTasksManager must be used within a TaskManager provider');
    }

    return context;
};
