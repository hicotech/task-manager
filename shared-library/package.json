{
    "name": "@task-manager/shared-library",
    "version": "1.0.0",
    "types": "dist/index.d.ts",
    "author": "Stanislav Majlath",
    "private": true,
    "scripts": {
        "build": "rm -rf dist && tsc src/types/index.d.ts --declaration --emitDeclarationOnly --outDir dist && cp src/types/index.d.ts dist"
    }
}