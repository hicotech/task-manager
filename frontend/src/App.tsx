import type { Task, TaskList } from '@task-manager/shared-library';
import { useEffect, useRef, useState } from 'react';
import { styled } from 'styled-components';
import CompletedTasks from './components/CompletedTasks/CompletedTasks';
import EmptyState from './components/EmptyState/EmptyState';
import Tasks from './components/Tasks/Tasks';
import { VITE_BACKEND_HOST, VITE_TASKS_ENDPOINT } from './constants';
import { completeTask, setTasks } from './context/actions/tasksManager';
import { useTasksManager } from './context/TaskManager';

const Container = styled.div`
    display: flex;
    justify-content: center;
    width: 100%;
    height: 100%;
`;

const ContentContainer = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: 0.5rem;
    width: 100%;
    height: 100%;

    &:has(:nth-child(n + 2)) {
        margin-top: 1.375rem;
        justify-content: initial;
    }
`;

const App = () => {
    // Use Abort Controller to abort any hanging API calls on new request
    const abortController = useRef<AbortController>(null);

    // Simple local state management with useState for TaskList name
    // Eventually could be implemented into Task Manager context
    const [tasksListName, setTasksListName] = useState<TaskList['name']>();

    // Globally available Task Manager
    const { state, dispatch } = useTasksManager();

    // Use ENVIRONMENT variables for Host and Endpoint data
    const url = `${VITE_BACKEND_HOST}/api/${VITE_TASKS_ENDPOINT}`;

    /**
     * Handle fetching data from backend and abort any hanging calls when executed
     *
     * Prepare TaskManagerAction object and pass it to Task Manager dispatch function which will set list of incomplete Tasks in global context state
     * Set TaskList name to local state
     */
    const fetchData = async () => {
        if (abortController.current) {
            abortController.current.abort();
        }

        abortController.current = new AbortController();

        try {
            const response = await fetch(url, {
                method: 'GET',
                signal: abortController.current.signal,
            });
            const result = await response.json();

            setTasksListName(result.name);
            dispatch(setTasks(result.tasks));
        } catch (error) {
            throw new Error(`Error fetching data: ${error}`);
        }
    };

    /**
     * Handle onCompleteTask callback
     *
     * Prepare TaskManagerAction object and pass it to Task Manager dispatch function which will set Task as completed in global context state
     *
     * @param taskId UUID of Task to be set as completed
     */
    const onCompleteTask = (taskId: Task['id']) => {
        dispatch(completeTask(taskId));
    };

    // Component unmount cleanup
    useEffect(() => {
        return () => {
            abortController.current?.abort();
        };
    }, []);

    return (
        <Container>
            {tasksListName && Boolean(state.tasks.length) ? (
                <Tasks
                    name={tasksListName}
                    tasks={state.tasks}
                    onButtonClick={onCompleteTask}
                    data-testid="tasks-list"
                />
            ) : (
                <ContentContainer>
                    <EmptyState onRequestTasks={fetchData} data-testid="empty-state" />
                    {Boolean(state.completed.length) && (
                        <CompletedTasks tasks={state.completed} data-testid="completed-list" />
                    )}
                </ContentContainer>
            )}
        </Container>
    );
};

export default App;
