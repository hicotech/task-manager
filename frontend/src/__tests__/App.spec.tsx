import { TaskList } from '@task-manager/shared-library';
import { render } from '@testing-library/react';
import React from 'react';
import App from '../App';
import * as TaskManagerContext from '../context/TaskManager';

jest.mock('../constants', () => ({
    VITE_BACKEND_HOST: 'test-backend-host',
    VITE_TASKS_ENDPOINT: 'test-tasks-endpoint',
}));

describe('App', () => {
    const testData: TaskList = {
        name: 'tasks-name',
        tasks: [
            {
                id: 'task-1',
                name: 'Task Circle',
                type: 'circle',
            },
            {
                id: 'task-2',
                name: 'Task Rectangle',
                type: 'rectangle',
            },
            {
                id: 'task-3',
                name: 'Task Polygon',
                type: 'polygon',
            },
        ],
    };

    afterEach(() => {
        jest.restoreAllMocks();
    });

    it('Renders Default', () => {
        const { container, queryByTestId } = render(<App />);

        expect(container).toMatchSnapshot();

        expect(queryByTestId('empty-state')).toBeInTheDocument();
        expect(queryByTestId('tasks-list')).not.toBeInTheDocument();
        expect(queryByTestId('completed-list')).not.toBeInTheDocument();
    });

    describe('Renders Tasks List', () => {
        it('When Completed List is empty', () => {
            // Mock local state value only once to prevent unwanted changes to other useState hooks
            jest.spyOn(React, 'useState').mockImplementationOnce(() => [testData.name, jest.fn()]);

            // Mock context state value
            jest.spyOn(TaskManagerContext, 'useTasksManager').mockImplementation(() => ({
                state: {
                    tasks: testData.tasks,
                    completed: [],
                },
                dispatch: jest.fn(),
            }));

            const { container, queryByTestId } = render(<App />);

            expect(container).toMatchSnapshot();

            expect(queryByTestId('empty-state')).not.toBeInTheDocument();
            expect(queryByTestId('tasks-list')).toBeInTheDocument();
            expect(queryByTestId('completed-list')).not.toBeInTheDocument();
        });

        it('When Completed List is not empty', () => {
            // Mock local state value only once to prevent unwanted changes to other useState hooks
            jest.spyOn(React, 'useState').mockImplementationOnce(() => [testData.name, jest.fn()]);

            // Mock context state value
            jest.spyOn(TaskManagerContext, 'useTasksManager').mockImplementation(() => ({
                state: {
                    tasks: testData.tasks.slice(0, 2),
                    completed: testData.tasks.slice(2, 3),
                },
                dispatch: jest.fn(),
            }));

            const { container, queryByTestId } = render(<App />);

            expect(container).toMatchSnapshot();

            expect(queryByTestId('empty-state')).not.toBeInTheDocument();
            expect(queryByTestId('tasks-list')).toBeInTheDocument();
            expect(queryByTestId('completed-list')).not.toBeInTheDocument();
        });
    });

    it('Renders Empty State and Completed List', () => {
        // Mock context state value
        jest.spyOn(TaskManagerContext, 'useTasksManager').mockImplementation(() => ({
            state: {
                tasks: [],
                completed: testData.tasks,
            },
            dispatch: jest.fn(),
        }));

        const { container, queryByTestId } = render(<App />);

        expect(container).toMatchSnapshot();

        expect(queryByTestId('empty-state')).toBeInTheDocument();
        expect(queryByTestId('tasks-list')).not.toBeInTheDocument();
        expect(queryByTestId('completed-list')).toBeInTheDocument();
    });
});
