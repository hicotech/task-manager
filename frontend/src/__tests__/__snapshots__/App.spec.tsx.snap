// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders Default 1`] = `
.c2 {
  display: flex;
  align-items: center;
}

.c3 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #52778b;
  background-color: transparent;
  border-radius: 100rem;
  border: 2px solid #52778b;
  padding: 0.375rem 1.25rem;
}

.c3:hover {
  color: #ffffff;
  background-color: #52778b;
}

.c3:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #52778b;
}

.c0 {
  display: flex;
  justify-content: center;
  width: 100%;
  height: 100%;
}

.c1 {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 0.5rem;
  width: 100%;
  height: 100%;
}

.c1:has(:nth-child(n + 2)) {
  margin-top: 1.375rem;
  justify-content: initial;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        data-testid="empty-state"
      >
        <button
          class="c3"
        >
          Requests Tasks
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`App Renders Empty State and Completed List 1`] = `
.c4 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c5 {
  text-align: center;
}

.c6 {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c7 {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 1px dashed #335a6f;
}

.c2 {
  display: flex;
  align-items: center;
}

.c3 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #52778b;
  background-color: transparent;
  border-radius: 100rem;
  border: 2px solid #52778b;
  padding: 0.375rem 1.25rem;
}

.c3:hover {
  color: #ffffff;
  background-color: #52778b;
}

.c3:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #52778b;
}

.c0 {
  display: flex;
  justify-content: center;
  width: 100%;
  height: 100%;
}

.c1 {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 0.5rem;
  width: 100%;
  height: 100%;
}

.c1:has(:nth-child(n + 2)) {
  margin-top: 1.375rem;
  justify-content: initial;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        data-testid="empty-state"
      >
        <button
          class="c3"
        >
          Requests Tasks
        </button>
      </div>
      <div
        class="c4"
        data-testid="completed-list"
      >
        <h1
          class="c5"
        >
          Completed Tasks
        </h1>
        <ul
          class="c6"
        >
          <li
            class="c7"
            id="task-1"
          >
            <svg
              data-testid="icon-circle"
              fill="#82baa4"
              height="1.5rem"
              viewBox="0 0 24 24"
              width="1.5rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
            </svg>
            Task Circle
          </li>
          <li
            class="c7"
            id="task-2"
          >
            <svg
              data-testid="icon-rectangle"
              fill="#6491a9"
              height="1.5rem"
              viewBox="0 0 24 24"
              width="1.5rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="24"
                width="24"
              />
            </svg>
            Task Rectangle
          </li>
          <li
            class="c7"
            id="task-3"
          >
            <svg
              data-testid="icon-polygon"
              fill="#bac1c4"
              height="1.5rem"
              viewBox="0 0 24 24"
              width="1.5rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="12,0 24,8 19,24 4,24 0,8"
              />
            </svg>
            Task Polygon
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`App Renders Tasks List When Completed List is empty 1`] = `
.c1 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.c2 {
  text-align: center;
}

.c3 {
  width: 100%;
}

.c4 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c5 {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem;
  border-radius: 1rem;
  border: 1px dashed #335a6f;
}

.c5:nth-child(even) {
  background-color: #e6e6e6;
}

.c6 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #528b7b;
  background-color: #ffffff;
  border-radius: 100rem;
  border: 2px solid #528b7b;
  padding: 0.375rem 1.25rem;
}

.c6:hover {
  color: #ffffff;
  background-color: #528b7b;
}

.c6:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #528b7b;
}

.c0 {
  display: flex;
  justify-content: center;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="tasks-list"
    >
      <h1
        class="c2"
      >
        tasks-name
      </h1>
      <progress
        class="c3"
        value="0"
      />
      <ul
        class="c4"
      >
        <li
          class="c5"
        >
          Task Circle
          <button
            class="c6"
          >
            Complete Task
          </button>
        </li>
        <li
          class="c5"
        >
          Task Rectangle
          <button
            class="c6"
          >
            Complete Task
          </button>
        </li>
        <li
          class="c5"
        >
          Task Polygon
          <button
            class="c6"
          >
            Complete Task
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`App Renders Tasks List When Completed List is not empty 1`] = `
.c1 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.c2 {
  text-align: center;
}

.c3 {
  width: 100%;
}

.c4 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c5 {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem;
  border-radius: 1rem;
  border: 1px dashed #335a6f;
}

.c5:nth-child(even) {
  background-color: #e6e6e6;
}

.c6 {
  display: inline-block;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: #528b7b;
  background-color: #ffffff;
  border-radius: 100rem;
  border: 2px solid #528b7b;
  padding: 0.375rem 1.25rem;
}

.c6:hover {
  color: #ffffff;
  background-color: #528b7b;
}

.c6:active {
  border: 2px solid #ffffff;
  box-shadow: 0px 0px 0px 2px #528b7b;
}

.c0 {
  display: flex;
  justify-content: center;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="tasks-list"
    >
      <h1
        class="c2"
      >
        tasks-name
      </h1>
      <progress
        class="c3"
        value="0"
      />
      <ul
        class="c4"
      >
        <li
          class="c5"
        >
          Task Circle
          <button
            class="c6"
          >
            Complete Task
          </button>
        </li>
        <li
          class="c5"
        >
          Task Rectangle
          <button
            class="c6"
          >
            Complete Task
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
